### PR TITLE
Do not use Vagrant for CentOS 7/8

### PR DESCRIPTION
**Pull request by adrianreber**
As Cirrus CI does not provide a real terminal this uses the same 'ssh -tt' workaround as the Vagrant setup. This sets up the CentOS 7 and 8 to allow SSH as root to localhost so that we can run all the tests via 'ssh -tt'.

Not going through vagrant reduces CI times for CentOS 7 and 8 from 6 minutes to 4 minutes.

### CODE REVIEW

File: .cirrus.yml [hunk 1]
```diff
@@ -65,3 +66,130 @@ vagrant_task:
...
+  integration_systemd_rootless_script: |
+    echo "SKIP: integration_systemd_rootless_script requires cgroup v2"
+  integration_fs_rootless_script: |
+    echo "SKIP: FIXME: integration_fs_rootless_script is skipped because of EPERM on writing cgroup.procs"
+
+task:
+  name: ci / centos 8
+  compute_engine_instance:
+    image_project: centos-cloud
+    image: family/centos-8
```
Comment (AkihiroSuda) [inline .cirrus.yml:139]: I'm not sure we want CentOS 8 CI.
CentOS 8 is going to die soon.

File: .cirrus.yml [hunk 2]
```diff
@@ -65,3 +66,130 @@ vagrant_task:
...
+    memory: 8
+
+  timeout_in: 30m
+  env:
+    HOME: /root
+    CIRRUS_WORKING_DIR: /home/runc
+    GO_VERSION: "1.16.4"
+    BATS_VERSION: "v1.3.0"
+
+  install_dependencies_script: |
```
Comment (AkihiroSuda) [inline .cirrus.yml:86]: Can we now remove `Vagrantfile.centos7`?

File: .cirrus.yml (hunk 1 above)
Comment (adrianreber) [reply .cirrus.yml]: replaced centos-8 with centos-stream-8. I can remove the complete centos 8 section if unnecessary. Just thought I add it.

File: .cirrus.yml (hunk 2 above)
Comment (adrianreber) [reply .cirrus.yml:86]: removed

File: .cirrus.yml [hunk 3]
```diff
@@ -65,3 +66,130 @@ vagrant_task:
...
+  unit_tests_script: |
+    ssh -tt localhost "make -C /home/runc localunittest"
+  integration_systemd_script: |
+    ssh -tt localhost "make -C /home/runc localintegration RUNC_USE_SYSTEMD=yes"
+  integration_fs_script: |
+    ssh -tt localhost "make -C /home/runc localintegration"
+  integration_systemd_rootless_script: |
+    echo "SKIP: integration_systemd_rootless_script requires cgroup v2"
+  integration_fs_rootless_script: |
+    echo "SKIP: FIXME: integration_fs_rootless_script is skipped because of EPERM on writing cgroup.procs"
```
Comment (AkihiroSuda) [inline .cirrus.yml:195]: I guess this should not be skipped for CentOS 8.

Also, can we use `matrix` to reduce code clone?

Comment (adrianreber) [reply .cirrus.yml:195]: > I guess this should not be skipped for CentOS Stream 8.

I tried to make it work for a couple of hours but it always fails with:
```
not ok 117 update rt period and runtime
# (in test file tests/integration/update.bats, line 548)
#   `echo "$root_period" >"$target_period"' failed
# runc spec --rootless (status=0):
# 
# Writing 1000000 to /sys/fs/cgroup/cpu,cpuacct//cpu.rt_period_us
# /tmp/bats-run-106184/bats.115768.src: line 548: /sys/fs/cgroup/cpu,cpuacct//cpu.rt_period_us: Permission denied
```

> Also, can we use `matrix` to reduce code clone?

Yes. I thought about it, but was unsure what would be better. I will update the PR.

File: .cirrus.yml (hunk 1 above)
Comment (kolyshkin) [reply .cirrus.yml]: I think it makes sense to have it (and I always wanted to add it).

CentOS stream 8 makes sense 👍🏻

File: .cirrus.yml [hunk 4]
```diff
@@ -65,3 +66,79 @@ vagrant_task:
     else
       ssh -tt default "sudo -i make -C /vagrant localrootlessintegration"
     fi
+
+task:
+  timeout_in: 30m
+  env:
+    HOME: /root
+    CIRRUS_WORKING_DIR: /home/runc
+    GO_VERSION: "1.16.4"
```
Comment (kolyshkin) [inline .cirrus.yml:75]: Can switch to 1.16.6 while at it.

File: .cirrus.yml [hunk 5]
```diff
@@ -65,3 +66,79 @@ vagrant_task:
...
+    git clone https://github.com/bats-core/bats-core
+    cd bats-core
+    git checkout $BATS_VERSION
+    ./install.sh /usr/local
+    cd -
+    if [ $DISTRO == centos-7 ]; then
+      # sysctl
+      echo "user.max_user_namespaces=15076" > /etc/sysctl.d/userns.conf
+      sysctl --system
+    fi
```
Comment (kolyshkin) [inline .cirrus.yml:112]: Move it to under the previous if / switch

File: .cirrus.yml (hunk 3 above)
Comment (kolyshkin) [reply .cirrus.yml:144]: > /tmp/bats-run-106184/bats.115768.src: line 548: /sys/fs/cgroup/cpu,cpuacct//cpu.rt_period_us: Permission denied

Mea culpa. Fix is in https://github.com/opencontainers/runc/pull/3106 -- @adrianreber you can cherry-pick it to this PR.

File: .cirrus.yml [hunk 6]
```diff
@@ -65,3 +66,79 @@ vagrant_task:
...
+      DISTRO: centos-7
+      DISTRO: centos-stream-8
+
+  name: ci / $DISTRO
+  compute_engine_instance:
+    image_project: centos-cloud
+    image: family/$DISTRO
+    platform: linux
+    cpu: 4
+    memory: 8
```
Comment (kolyshkin) [inline .cirrus.yml:88]: Can you use a suffix here? I.e. `8G`.

File: .cirrus.yml [hunk 7]
```diff
@@ -65,3 +66,79 @@ vagrant_task:
...
+
+  install_dependencies_script: |
+    if [ $DISTRO == centos-7 ]; then
```
Comment (kolyshkin) [inline .cirrus.yml:91]:
```bash
case $DISTRO in
centos-7)
    .....
    ;;
centos-stream-8)
    ....
    ;;
esac
```

File: .cirrus.yml [hunk 8]
```diff
@@ -65,3 +66,79 @@ vagrant_task:
...
+      yum install -y -q epel-release
+      (cd /etc/yum.repos.d && curl -O https://copr.fedorainfracloud.org/coprs/adrian/criu-el7/repo/epel-7/adrian-criu-el7-epel-7.repo)
+    elif [ $DISTRO == centos-stream-8 ]; then
+      yum install -y -q https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm dnf-plugins-core
```
Comment (kolyshkin) [inline .cirrus.yml:95]: It seems that epel-release is available in centos-stream-8 repo (in extras): http://mirror.centos.org/centos/8-stream/extras/x86_64/os/Packages/

So maybe
```bash
yum --enablerepo extras install -y -q epel-release dnf-plugins-core
```

**Comment by kolyshkin** [comment]
CI failure in Fedora 34 ("not ok 18 checkpoint --lazy-pages and restore") may be a flake I am fighting for quite some time (see https://github.com/opencontainers/runc/issues/2924). It is rare so for now the workaround is to restart CI. It might have been fixed with criu 1.16 (https://github.com/checkpoint-restore/criu/issues/1524) but more changes are needed from runc side (I have a WIP PR, https://github.com/opencontainers/runc/pull/2761, that needs a lot of cleaning), and I have yet to re-check that.

File: .cirrus.yml [hunk 9]
```diff
@@ -1,29 +1,30 @@
...
-  image_project: cirrus-images
-  image: family/docker-kvm
-  platform: linux
-  nested_virtualization: true
-  # CPU limit: `16 / NTASK`: see https://cirrus-ci.org/faq/#are-there-any-limits
-  cpu: 8
-  # Memory limit: `4GB * NCPU`
-  memory: 32G
+task:
+  name: ci / fedora 34 (vagrant)
```
Comment (kolyshkin) [inline .cirrus.yml:10]: Can be not rename the test (it is currently called `vagrant DISTRO:fedora34`)? It is marked as required in this repo configuration, and we'll need to change that.

Or, if you're renaming it, maybe call it vagrant/fedora or something, so we would not have to change it when we switch to f35.

File: .cirrus.yml (hunk 3 above)
Comment (adrianreber) [reply .cirrus.yml]: With your patch from #3106 it fails with:
```
# (in test file tests/integration/update.bats, line 548)
# Writing 1000000 to /sys/fs/cgroup/cpu,cpuacct/runc-cgroups-integration-test/cpu.rt_period_us
# /tmp/bats-run-106836/bats.116418.src: line 548: /sys/fs/cgroup/cpu,cpuacct/runc-cgroups-integration-test/cpu.rt_period_us: Permission denied
```

File: .cirrus.yml (hunk 4 above)
Comment (adrianreber) [reply .cirrus.yml:79]: done

File: .cirrus.yml (hunk 5 above)
Comment (adrianreber) [reply .cirrus.yml]: done

File: .cirrus.yml (hunk 6 above)
Comment (adrianreber) [reply .cirrus.yml]: done

File: .cirrus.yml (hunk 9 above)
Comment (adrianreber) [reply .cirrus.yml]: Changed the name back to `vagrant DISTRO:fedora34`. I still kept the new names of the CentOS tests as they are no longer vagrant based.

File: .cirrus.yml (hunk 3 above)
Comment (kolyshkin) [reply .cirrus.yml]: Yeah, this makes more sense. I'm afraid we need to have a hack for this in `tests/rootless.sh`'s `enable_cgroup`. Something like this:

```diff
diff --git a/tests/rootless.sh b/tests/rootless.sh
index bacea49d..36db5340 100755
--- a/tests/rootless.sh
+++ b/tests/rootless.sh
@@ -114,6 +114,8 @@ function enable_cgroup() {
                # necessary, and might actually be a bug in our impl of cgroup
                # handling.
                [[ "$cg" == "cpuset" ]] && chown rootless:rootless "$CGROUP_MOUNT/$cg$CGROUP_PATH/cpuset."{cpus,mems}
+               # The following is required by "update rt period and runtime".
+               [[ "$cg" == "cpu" ]] && chown rootless:rootless "$CGROUP_MOUNT/$cg$CGROUP_PATH/cpu.rt_"{period,quota}us
        done
        # cgroup v2
        if [[ -e "$CGROUP_MOUNT/cgroup.controllers" ]]; then
```

@AkihiroSuda do you maybe have some input on this?

Comment (kolyshkin) [reply .cirrus.yml]: I am not sure this test ever worked. On GHA (ubuntu, rootless_cgroups) it says

> ok 118 update rt period and runtime # skip test requires cgroups_rt

File: .cirrus.yml (hunk 4 above)
Comment (kolyshkin) [reply .cirrus.yml:79]: Looks like not.

Comment (adrianreber) [reply .cirrus.yml]: I was sure I did that change. Second try pushed just no.w

File: .cirrus.yml (hunk 3 above)
Comment (kolyshkin) [reply .cirrus.yml]: I have added the above fix to https://github.com/opencontainers/runc/pull/3106 -- again, feel free to cherry-pick @adrianreber

Comment (adrianreber) [reply .cirrus.yml]: The added fix to #3106 makes other CI runs fail. As it can be seen in #3106

Comment (adrianreber) [reply .cirrus.yml]: I adapted your second patch and now everything seems happy.

**Review by kolyshkin — APPROVED** [review]
LGTM

**Review by AkihiroSuda — APPROVED** [review]

**Review by mrunalp — APPROVED** [review]

**Pull request merged**